velox/expression/signature_parser/SignatureParser.yy

```diff
@@ -0,0 +1,98 @@
+%{
+#include <FlexLexer.h>
+#include "velox/common/base/Exceptions.h"
+#include "velox/expression/TypeSignature.h"
+#include "velox/expression/signature_parser/ParseUtil.h"
+%}
+%require "3.0.4"
+%language "C++"
+
+%define parser_class_name {Parser}
+%define api.namespace {facebook::velox::exec}
+%define api.value.type variant
+%parse-param {Scanner* scanner}
+%define parse.error verbose
+
+%code requires
+{
+    namespace facebook::velox::exec {
+        class Scanner;
+        class TypeSignature;
+    } // namespace facebook::velox::exec
+} // %code requires
+
+%code
+{
+    #include <velox/expression/signature_parser/Scanner.h>
+    #define yylex(x) scanner->lex(x)
+}
+
+%token               LPAREN RPAREN COMMA ARRAY MAP ROW FUNCTION DECIMAL
+%token <std::string> WORD VARIABLE QUOTED_ID
+%token YYEOF         0
+
+%nterm <std::shared_ptr<exec::TypeSignature>> special_type function_type decimal_type row_type array_type map_type
+%nterm <std::shared_ptr<exec::TypeSignature>> type named_type
+%nterm <std::vector<exec::TypeSignature>> type_list type_list_opt_names
+%nterm <std::vector<std::string>> type_with_spaces
+
+%%
+
+type_spec : type                 { scanner->setTypeSignature($1); }
```

Comment:
it's a bummer we need to repeat the type grammar here. Is there no way refactor and include that part?

Reply:
Each rule returns a TypeSignature here vs. a TypePtr in the Type Parser.
To unify both, we have to make each rule return strings and finally return a parse tree of strings. Then each parser has to recurse the tree again to convert the string to a TypeSignature or TypePtr.
I am planning to use the TypeParser in Prestissimo to replace Antlr [here](https://github.com/prestodb/presto/pull/21357). 
I felt the double-pass approach would be inefficient.

Reply:
Ok, thanks for looking into this!